Show note where the macro failed to match

Noratrieb opened the pull request:
When feeding the wrong tokens, it used to fail with a very generic error that wasn't very helpful. This change tries to help by noting where specifically the matching went wrong.

```rust
macro_rules! uwu {
    (a a a b) => {};
}
uwu! { a a a c }
```

```diff
error: no rules expected the token `c`
 --> macros.rs:5:14
  |
1 | macro_rules! uwu {
  | ---------------- when calling this macro
...
4 | uwu! { a a a c }
  |              ^ no rules expected this token in macro call
  |
+note: while trying to match `b`
+ --> macros.rs:2:12
+  |
+2 |     (a a a b) => {};
+  |            ^
```

rust-highfive commented:
r? @lcnr

(rust-highfive has picked a reviewer for you, use r? to override)

Dylan-DPC commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 03254db9f4cfd213bb6ba05fa435097b51aabe89 with merge 1b60c6801809977dd5201067b34e8f2ffb884eb0...
<!-- homu: {"type":"TryBuildStarted","head_sha":"03254db9f4cfd213bb6ba05fa435097b51aabe89","merge_sha":"1b60c6801809977dd5201067b34e8f2ffb884eb0"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3307621293/jobs/5459965305)
Build commit: 1b60c6801809977dd5201067b34e8f2ffb884eb0 (`1b60c6801809977dd5201067b34e8f2ffb884eb0`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3307621293/jobs/5459965305"},"merge_sha":"1b60c6801809977dd5201067b34e8f2ffb884eb0"} -->

rust-timer commented:
Queued 1b60c6801809977dd5201067b34e8f2ffb884eb0 with parent 9be2f35a4c1ed1b04aa4a6945b64763f599259ff, future [comparison URL](https://perf.rust-lang.org/compare.html?start=9be2f35a4c1ed1b04aa4a6945b64763f599259ff&end=1b60c6801809977dd5201067b34e8f2ffb884eb0).

rust-timer commented:
Finished benchmarking commit (1b60c6801809977dd5201067b34e8f2ffb884eb0): [comparison URL](https://perf.rust-lang.org/compare.html?start=9be2f35a4c1ed1b04aa4a6945b64763f599259ff&end=1b60c6801809977dd5201067b34e8f2ffb884eb0&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=9be2f35a4c1ed1b04aa4a6945b64763f599259ff&end=1b60c6801809977dd5201067b34e8f2ffb884eb0&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 2.0%     | [0.2%, 7.0%] | 31        |
| Regressions ❌ <br /> (secondary) | 2.2%     | [0.6%, 4.9%] | 18        |
| Improvements ✅ <br /> (primary) | -0.6%    | [-0.6%, -0.6%] | 1         |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | 1.9%     | [-0.6%, 7.0%] | 32        |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=9be2f35a4c1ed1b04aa4a6945b64763f599259ff&end=1b60c6801809977dd5201067b34e8f2ffb884eb0&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | -        | -     | 0         |
| Regressions ❌ <br /> (secondary) | -        | -     | 0         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -5.0%    | [-5.0%, -5.0%] | 1         |
| All ❌✅ (primary) | -        | -     | 0         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=9be2f35a4c1ed1b04aa4a6945b64763f599259ff&end=1b60c6801809977dd5201067b34e8f2ffb884eb0&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 4.4%     | [2.1%, 6.4%] | 8         |
| Regressions ❌ <br /> (secondary) | 3.3%     | [2.1%, 5.1%] | 10        |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | 4.4%     | [2.1%, 6.4%] | 8         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

Nilstrieb commented:
Yes, as expected, the naive implementation is a regression as this path is very hot. I'll reimplement the diagnostic in a different way that keeps all new code in the cold diagnostics path.

jruderman commented on src/test/ui/macros/macro-non-lifetime.stderr:
```diff
@@ -6,6 +6,12 @@ LL | macro_rules! m { ($x:lifetime) => { } }
 LL |     m!(a);
    |        ^ no rules expected this token in macro call
+   |
+note: while trying to match `meta-variable $x:lifetime`
```
Should the word "meta-variable" be outside the code-quotes?

Nilstrieb replied:
yeah the quoting is quite bad i need to fix it

lcnr commented:
r? compiler

Nilstrieb commented:
r? @ghost

this needs some more work and won't be ready for review for a while

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3316264953/jobs/5477857064) [(plain)](https://github.com/rust-lang/rust/commit/277bcdad0ffbc5d284b107575d0e281ecb30a9cc/checks/9079746931/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/compiler/rustc_expand/src/mbe/macro_rules.rs at line 381:
             Failure(token, msg) => match self.best_failure {
                 Some((ref best_token, _, _)) if best_token.span.lo() >= token.span.lo() => {}
-                    self.best_failure =
-                    self.best_failure =
-                        Some((token.clone(), msg, self.remaining_matcher.expect("must have collected matcher already").clone()))
+                    self.best_failure = Some((
+                        msg,
+                        self.remaining_matcher
+                        self.remaining_matcher
+                            .expect("must have collected matcher already")
+                            .clone(),
                 }
             },
             },
             Error(err_sp, msg) => {
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_expand/src/mbe/macro_rules.rs" "/checkout/compiler/rustc_expand/src/mbe/transcribe.rs" "/checkout/compiler/rustc_expand/src/mbe/metavar_expr.rs" "/checkout/compiler/rustc_expand/src/mut_visit/tests.rs" "/checkout/compiler/rustc_expand/src/tokenstream/tests.rs" "/checkout/compiler/rustc_expand/src/mbe/quoted.rs" "/checkout/compiler/rustc_expand/src/proc_macro.rs" "/checkout/compiler/rustc_expand/src/errors.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

compiler-errors commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit a9d7d01d271df3270755e871c261e377766e2f8a with merge 0ca851de666e863a0e2d14d192598cabc510b640...
<!-- homu: {"type":"TryBuildStarted","head_sha":"a9d7d01d271df3270755e871c261e377766e2f8a","merge_sha":"0ca851de666e863a0e2d14d192598cabc510b640"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3340602354/jobs/5532643091)
Build commit: 0ca851de666e863a0e2d14d192598cabc510b640 (`0ca851de666e863a0e2d14d192598cabc510b640`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3340602354/jobs/5532643091"},"merge_sha":"0ca851de666e863a0e2d14d192598cabc510b640"} -->

rust-timer commented:
Queued 0ca851de666e863a0e2d14d192598cabc510b640 with parent 0da281b6068a7d889ae89a9bd8991284cc9b7535, future [comparison URL](https://perf.rust-lang.org/compare.html?start=0da281b6068a7d889ae89a9bd8991284cc9b7535&end=0ca851de666e863a0e2d14d192598cabc510b640).

rust-timer commented:
Finished benchmarking commit (0ca851de666e863a0e2d14d192598cabc510b640): [comparison URL](https://perf.rust-lang.org/compare.html?start=0da281b6068a7d889ae89a9bd8991284cc9b7535&end=0ca851de666e863a0e2d14d192598cabc510b640&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=0da281b6068a7d889ae89a9bd8991284cc9b7535&end=0ca851de666e863a0e2d14d192598cabc510b640&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | -        | -     | 0         |
| Regressions ❌ <br /> (secondary) | 0.2%     | [0.2%, 0.2%] | 1         |
| Improvements ✅ <br /> (primary) | -0.9%    | [-1.4%, -0.4%] | 11        |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | -0.9%    | [-1.4%, -0.4%] | 11        |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=0da281b6068a7d889ae89a9bd8991284cc9b7535&end=0ca851de666e863a0e2d14d192598cabc510b640&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 1.5%     | [1.5%, 1.5%] | 1         |
| Regressions ❌ <br /> (secondary) | -        | -     | 0         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -3.3%    | [-4.8%, -1.8%] | 2         |
| All ❌✅ (primary) | 1.5%     | [1.5%, 1.5%] | 1         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=0da281b6068a7d889ae89a9bd8991284cc9b7535&end=0ca851de666e863a0e2d14d192598cabc510b640&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | -        | -     | 0         |
| Regressions ❌ <br /> (secondary) | -        | -     | 0         |
| Improvements ✅ <br /> (primary) | -2.0%    | [-2.3%, -1.6%] | 3         |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | -2.0%    | [-2.3%, -1.6%] | 3         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

compiler-errors commented:
Perf results look pretty good tbh.

Nilstrieb commented:
As expected, html5ever got faster, as we don't track the `best_failure` anymore on the happy path. deep-vector getting slower is a little concerning and I'll take a look at that later. But it's not bad.

Nilstrieb commented:
forgot about that, @rustbot ready

Nilstrieb commented:
Thinking about this again, I don't like the diagnostics here enough and I think they should be improved further. I do want to get the refactoring merged though, so I will split it into a seperate PR.
@rustbot author

Noratrieb committed:
Show a note where a macro failed to match

This shows a small note on what the macro matcher was currently
processing to aid with "no rules expected the token X" errors.

compiler-errors commented on uwu.rs:
```diff
@@ -0,0 +1,17 @@
+#![crate_type = "lib"]
```
<3

Nilstrieb replied:
<3

WaffleLapkin replied:
<3

BoxyUwU replied:
<3

asquared31415 replied:
<3

workingjubilee replied:
UwU

C0RR1T replied:
![2Q==.jpg](https://user-images.githubusercontent.com/74185591/201497389-7b3b4ff9-9cbd-44be-818c-062efd6c19d1.jpg)

estebank commented:
@bors r+

bors commented:
:pushpin: Commit a34d1cfde326eec323e61f8701e0fac780b322e2 has been approved by `estebank`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=estebank a34d1cfde326eec323e61f8701e0fac780b322e2 -->
<!-- homu: {"type":"Approved","sha":"a34d1cfde326eec323e61f8701e0fac780b322e2","approver":"estebank","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Nilstrieb commented:
@estebank oh, I actually wanted to fix a few more diagnostics here before putting it up for review again but I've changed my mind on that, I think merging this is fine as is.. but can you `r-` it so I at least fix the commit message? :)

compiler-errors commented:
@bors r-
@rustbot author

yeet

bors commented:
:pushpin: Commit 7e7c11cf560527c4a7c70da63455cc23f4c51235 has been approved by `estebank`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=estebank 7e7c11cf560527c4a7c70da63455cc23f4c51235 -->
<!-- homu: {"type":"Approved","sha":"7e7c11cf560527c4a7c70da63455cc23f4c51235","approver":"estebank","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

C0RR1T commented:
Hyped for this one
